File: CPP/Algorithm/Maximum_Xor_Subarray.cpp
// Complexity: O(n)
// Application: Used to determine the maximum xor subarray using trie.
// const long long N = 6e6 + 5;  // n * 30

long long trie[N][2], c = 0;
void insert(auto x) {
    long long nd = 0;
    for (int i = 30; i >= 0; --i) {
        long long y = x >> i & 1;
        if (!trie[nd][y]) trie[nd][y] = ++c;
        nd = trie[nd][y];
    }
}
long long search(auto x) {
    long long ans = 0, nd = 0;
    for (int i = 30; i >= 0; --i) {
        long long y = x >> i & 1;
        if (trie[nd][y ^ 1]) {
            nd = trie[nd][y ^ 1]; ans += (1 << i);
        }
        else nd = trie[nd][y];
    }
    return ans;
}

// Operation:
    cin >> n;
    long long xr = 0, mx = 0; insert(xr);
    for (i = 0; i < n; ++i) {
        long long x; cin >> x; xr ^= x; insert(xr);
        mx = max(search(xr), mx);
    }
    cout << mx << '\n';
    for (i = 0; i <= n; ++i) trie[i][0] = trie[i][1] = 0;

File: Python/Algorithm/Maximum_Xor_Subarray.py
# Complexity: O(n)
# Application: Used to determine the maximum xor sublsay using trie.

class TrieNode:
    def __init__(self): self.children = {}

def insert(root, x):
    node = root
    for i in range(30, -1, -1):
        y = (x >> i) & 1
        if y not in node.children: node.children[y] = TrieNode()
        node = node.children[y]

def search(root, x):
    node, ans = root, 0
    for i in range(30, -1, -1):
        y = (x >> i) & 1
        if y ^ 1 in node.children:
            node = node.children[y ^ 1]; ans += (1 << i)
        else: node = node.children.get(y, node)
    return ans

def xor(ls):
    root = TrieNode(); xr, mx = 0, 0; insert(root, xr)
    for x in ls:
        xr ^= x; insert(root, xr)
        mx = max(mx, search(root, xr))
    return mx